# SPDX-FileCopyrightText: 2025 CoreWeave, Inc.
# SPDX-License-Identifier: GPL-2.0-or-later
# SPDX-PackageName: slurm-containers
diff --git a/slurm/slurm.h b/slurm/slurm.h
index a7e82f9..ee8bb91 100644
--- a/slurm/slurm.h
+++ b/slurm/slurm.h
@@ -2970,6 +2970,8 @@ typedef struct reservation_name_msg {
  */
 typedef struct {
 	time_t last_update;	/* last update time of the build parameters */
+	char *sunk_node_dashboard_url; /* sunk node grafana dashboard url */
+	char *sunk_job_dashboard_url; /* sunk job grafana dashboard url */
 	char *accounting_storage_tres; /* list of tres */
 	uint16_t accounting_storage_enforce; /* job requires valid association:
 					      * user/account/partition/cluster */
diff --git a/src/api/config_info.c b/src/api/config_info.c
index 0aeb413..8e2a1f8 100644
--- a/src/api/config_info.c
+++ b/src/api/config_info.c
@@ -739,6 +739,9 @@ extern void *slurm_ctl_conf_2_key_pairs(slurm_conf_t *conf)
 		xfree(key);
 	}

+	add_key_pair(ret_list, "SUNKNodeDashboardURL", "%s", conf->sunk_node_dashboard_url);
+	add_key_pair(ret_list, "SUNKJobDashboardURL", "%s", conf->sunk_job_dashboard_url);
+
 	if (xstrcmp(conf->priority_type, "priority/basic")) {
 		add_key_pair(ret_list, "FairShareDampeningFactor", "%u",
 			     conf->fs_dampening_factor);
diff --git a/src/api/node_info.c b/src/api/node_info.c
index d29a175..563868e 100644
--- a/src/api/node_info.c
+++ b/src/api/node_info.c
@@ -498,6 +498,15 @@ char *slurm_sprint_node_table(node_info_t *node_ptr, int one_liner)
 		xstrfmtcat(out, "Parameters=%s", node_ptr->parameters);
 	}

+	/****** Line (optional) ******/
+	if (slurm_conf.sunk_node_dashboard_url && node_ptr->comment) {
+		xstrcat(out, line_end);
+		// SUNK slurm node comment always contains the node name
+		xstrfmtcat(out, "SUNKNodeDashboardURL=%s?var-node=%s",
+			slurm_conf.sunk_node_dashboard_url,
+			node_ptr->comment);
+	}
+
 	if (one_liner)
 		xstrcat(out, "\n");
 	else
diff --git a/src/common/read_config.c b/src/common/read_config.c
index 382aaf6..4838906 100644
--- a/src/common/read_config.c
+++ b/src/common/read_config.c
@@ -308,6 +308,8 @@ s_p_options_t slurm_conf_options[] = {
 	{"LaunchType", S_P_STRING},
 	{"Licenses", S_P_STRING},
 	{"LogTimeFormat", S_P_STRING},
+	{"SUNKNodeDashboardURL", S_P_STRING},
+	{"SUNKJobDashboardURL", S_P_STRING},
 	{"MailDomain", S_P_STRING},
 	{"MailProg", S_P_STRING},
 	{"MaxArraySize", S_P_UINT32},
@@ -2708,6 +2710,8 @@ extern void free_slurm_conf(slurm_conf_t *ctl_conf_ptr, bool purge_node_hash)
 	xfree(ctl_conf_ptr->slurmd_spooldir);
 	xfree(ctl_conf_ptr->slurmd_user_name);
 	xfree(ctl_conf_ptr->srun_epilog);
+	xfree(ctl_conf_ptr->sunk_node_dashboard_url);
+	xfree(ctl_conf_ptr->sunk_job_dashboard_url);
 	xfree(ctl_conf_ptr->srun_port_range);
 	xfree(ctl_conf_ptr->srun_prolog);
 	xfree(ctl_conf_ptr->state_save_location);
@@ -2905,6 +2909,8 @@ void init_slurm_conf(slurm_conf_t *ctl_conf_ptr)
 	xfree(ctl_conf_ptr->state_save_location);
 	xfree(ctl_conf_ptr->suspend_exc_nodes);
 	xfree(ctl_conf_ptr->suspend_exc_parts);
+	xfree(ctl_conf_ptr->sunk_node_dashboard_url);
+	xfree(ctl_conf_ptr->sunk_job_dashboard_url);
 	xfree(ctl_conf_ptr->suspend_exc_states);
 	xfree(ctl_conf_ptr->suspend_program);
 	ctl_conf_ptr->suspend_rate = NO_VAL16;
@@ -5189,6 +5195,9 @@ static int _validate_and_set_defaults(slurm_conf_t *conf,
 	(void) s_p_get_string(&conf->task_epilog, "TaskEpilog", hashtbl);
 	(void) s_p_get_string(&conf->task_prolog, "TaskProlog", hashtbl);

+	(void) s_p_get_string(&conf->sunk_node_dashboard_url, "SUNKNodeDashboardURL", hashtbl);
+	(void) s_p_get_string(&conf->sunk_job_dashboard_url, "SUNKJobDashboardURL", hashtbl);
+
 	if (!s_p_get_uint16(&conf->tcp_timeout, "TCPTimeout", hashtbl))
 		conf->tcp_timeout = DEFAULT_TCP_TIMEOUT;

diff --git a/src/scontrol/info_job.c b/src/scontrol/info_job.c
index 8685b30..f9ca546 100644
--- a/src/scontrol/info_job.c
+++ b/src/scontrol/info_job.c
@@ -948,6 +948,15 @@ static char *_sprint_job_info(job_info_t *job_ptr)
 		xstrcat(out, line_end);
 	}

+	/****** Line (optional) ******/
+	if (slurm_conf.sunk_job_dashboard_url && slurm_conf.cluster_name) {
+		xstrfmtcat(out, "SUNKJobDashboardURL=%s?var-slurm_cluster=%s&var-job_id=%d",
+			slurm_conf.sunk_job_dashboard_url,
+			slurm_conf.cluster_name,
+			job_ptr->job_id);
+		xstrcat(out, line_end);
+	}
+
 	xstrtrim(out);

 	/****** END OF JOB RECORD ******/
diff --git a/src/slurmd/slurmd/req.c b/src/slurmd/slurmd/req.c
index c18c11c..66bbd50 100644
--- a/src/slurmd/slurmd/req.c
+++ b/src/slurmd/slurmd/req.c
@@ -2242,6 +2242,14 @@ static void _rpc_batch_job(slurm_msg_t *msg)

 	info("Launching batch %pI for UID %u", &req->step_id, batch_uid);

+	if (slurm_conf.sunk_job_dashboard_url && slurm_conf.cluster_name) {
+		info("Job %d: SUNK job dashboard: %s?var-slurm_cluster=%s&var-job_id=%d",
+			req->step_id.job_id,
+			slurm_conf.sunk_job_dashboard_url,
+			slurm_conf.cluster_name,
+			req->step_id.job_id);
+	}
+
 	debug3("%s: call to _forkexec_slurmstepd", __func__);
 	rc = _forkexec_slurmstepd(LAUNCH_BATCH_JOB, (void *)req, cli, batch_uid,
 				  req->step_id.job_id, SLURM_BATCH_SCRIPT,
@@ -4421,6 +4429,14 @@ static void _rpc_terminate_job(slurm_msg_t *msg)
 	int		delay;
 	bool send_response = true;

+	if (slurm_conf.sunk_job_dashboard_url && slurm_conf.cluster_name && req->step_id.job_id) {
+		info("Job %d: SUNK job dashboard: %s?var-slurm_cluster=%s&var-job_id=%d",
+			req->step_id.job_id,
+			slurm_conf.sunk_job_dashboard_url,
+			slurm_conf.cluster_name,
+			req->step_id.job_id);
+	}
+
 	debug("%s: starting for %pI %ps",
 	      __func__, &req->step_id, &req->step_id);

diff --git a/src/srun/launch.c b/src/srun/launch.c
index 42714f3..bdc47b6 100644
--- a/src/srun/launch.c
+++ b/src/srun/launch.c
@@ -1525,6 +1525,14 @@ extern int launch_g_step_launch(srun_job_t *job, slurm_step_io_fds_t *cio_fds,
 		}
 	}

+	if (slurm_conf.sunk_job_dashboard_url && slurm_conf.cluster_name && job->step_id.job_id) {
+		printf("Job %d: SUNK job dashboard: %s?var-slurm_cluster=%s&var-job_id=%d\n",
+			job->step_id.job_id,
+			slurm_conf.sunk_job_dashboard_url,
+			slurm_conf.cluster_name,
+			job->step_id.job_id);
+	}
+
 	update_job_state(job, SRUN_JOB_STARTING);
 	if (slurm_step_launch_wait_start(job->step_ctx) == SLURM_SUCCESS) {
 		update_job_state(job, SRUN_JOB_RUNNING);
diff --git a/src/srun/srun_job.c b/src/srun/srun_job.c
index 4c0f6c5..0884941 100644
--- a/src/srun/srun_job.c
+++ b/src/srun/srun_job.c
@@ -1592,6 +1592,14 @@ extern void fini_srun(srun_job_t *job, bool got_alloc, uint32_t *global_rc)
 		*global_rc = 128 + WTERMSIG(*global_rc);

 	mpir_cleanup();
+
+	if (slurm_conf.sunk_job_dashboard_url && slurm_conf.cluster_name && job->step_id.job_id) {
+		printf("Job %d: SUNK job dashboard: %s?var-slurm_cluster=%s&var-job_id=%d\n",
+			job->step_id.job_id,
+			slurm_conf.sunk_job_dashboard_url,
+			slurm_conf.cluster_name,
+			job->step_id.job_id);
+	}
 }

 void
